#534 Support Python 3

rmjarvis opened the pull request:
This PR converts the GalSim code to work with Python 3.  So if you have access to a Python 3 installation, please test that everything works for you there.  And if not, at least check that I didn't break your Python 2 installation with anything here.

I added a new dependency, called future, to help enable this.  While I didn't technically need it (I could have rolled my own versions of the features I use from there), it's a pretty easy dependency, so hopefully it won't be a significant hardship for anyone.  But it means you'll probably need to run `pip install future` before installing GalSim on this branch.

For reference, I used this handy [cheat sheet](http://python-future.org/compatible_idioms.html) extensively for writing the Python 2-3 compatible code.  This [guide](http://python3porting.com/noconv.html#noconv-chapter) also had some useful advice, but it was not nearly as comprehensive.

Finally, Python 3 added a ResourceWarning that lets you know when you've forgotten to close files.  I'm not sure if this only happens when running in nosetests or what (I didn't get them when I ran individual test files), but it helped me track down a lot of places where we were doing that.  The biggest culprit was `python.getdata`, which seems to not (or not always) close the input file.  So I switched all of those to `with python.open(..)` blocks, which fixed those errors.  There were a number of others that were our fault.  

The trickiest one was that `galsim.fits.read` could raise an exception before getting to the place where it closes the files.  It turns out, this is precisely what the python `try... finally` construct is for, so I used that, which makes for nicer code there.

rmandelb commented:
Installation broken on my Mac:
```
scons: Reading SConscript files ...
SCons is version 2.0.1 using python version 2.7.2
Python is from /Library/Frameworks/EPD64.framework/Versions/7.2/include/python2.7
Using the following (non-default) scons options:
   CXX = clang++
These can be edited directly in the file gs_scons.conf.
Type scons -h for a full list of available options.
Using python =  /Library/Frameworks/EPD64.framework/Versions/Current/bin/python
Using default PYPREFIX =  /Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/site-packages
Using compiler: /usr/bin/clang++
compiler version: 6.0
Determined that a good number of jobs = 8
Checking for C++ header file fftw3.h... (cached) yes
Checking for correct FFTW linkage... (cached) yes
Checking for boost header files... (cached) yes
Boost version is 1.49.0
Checking for C++ header file TMV.h... (cached) yes
TMV version is 0.72
Using TMV_LINK file: /usr/local/share/tmv/tmv-link
     -L/usr/local/lib -ltmv
Mac version is 10.9.5
XCode version is 6.1
Checking for correct TMV linkage... (this may take a little while)
Checking for correct TMV linkage... (cached) yes
Checking if we can build against Python... yes
Building for python version 2.7
Python libdir =  /Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/config
Python libfile =  libpython2.7.a
Checking if we can build module using TMV... yes
Checking if we can build against NumPy... 
Unable to import numpy using the python executable:
/Library/Frameworks/EPD64.framework/Versions/Current/bin/python
Mac version is 10.9.5
```

The relevant part of config.log is
```
scons: Configure: Checking if we can build against NumPy... 
.sconf_temp/conftest_68 <-
  |import numpy; print(numpy.get_include())
bash -c /Library/Frameworks/EPD64.framework/Versions/Current/bin/python < .sconf_temp/conftest_68 > .sconf_temp/conftest_68.out 2>&1
```

rmandelb commented on INSTALL.md:
```diff
@@ -62,6 +62,9 @@ The GalSim package also requires
   this package is happening.
 
+* The future module, which is used to ease compatibility between Python 2
+  and Python 3.
```
Elsewhere in INSTALL.md, there are strong statements about requiring python 2.6 or 2.7, so those also have to be updated to say that python 3 is required.

rmjarvis replied:
I think you mean permitted, not required.  But yes, good point.  I'll go through and check the language around those statements.

esheldon commented:
as I recall you support both astro.py fits reaing and fitsio.

with fitsio I would recommend using a `with` block to avoid the file closing problem, rather than finally, when it is possible to do so

```python
with fitsio.FITS(filename) as fits:
    # do something
```

rmjarvis commented:
> as I recall you support both astropy fits and fitsio.

GalSim doesn't actually use fitsio ever, just pyfits or astropy.  You're remembering TreeCorr where I supported both.  Although with the latest version I switching to just fitsio, since it's faster than astropy, and it's simple enough to have setup.py handle the dependency for people.  I haven't made the switch here for GalSim, although maybe someday...

> with fitsio I recommend using a with block to avoid the file closing problem, rather than finally, when it is possible to do so

We do use `with` blocks most of the time when we open files, although we had missed a couple, which I've now added.  In the `galsim.fits.py` routines, our desire to support reading from both a file (opening it for the user) or an open HDUList means that a `with` block doesn't really work (or at least I don't know how to make it work easily).  That's where the `finally` block was helpful.

rmandelb commented:
Well - this is weird.  I went to master, compiled, came back to #534, did "rm -rf .scon*", then compiled, and it didn't fail any more.  I don't understand this, but I'm proceeding bravely onwards anyhow...

I am getting a ton of warnings like this when compiling:
```
In file included from src/OneDimensionalDeviate.cpp:22:
In file included from include/galsim/OneDimensionalDeviate.h:26:
include/galsim/Random.h:44:32: warning: unknown warning group '-Wunused-local-typedefs', ignored
      [-Wunknown-pragmas]
#pragma GCC diagnostic ignored "-Wunused-local-typedefs"
```

rmjarvis commented:
I might have fixed the warnings for you.  And I moved all the warning suppression to a single place to make it easier to change things in the future.  (Should have done this earlier...)

rmandelb commented:
Yep, fixed.  Thanks.

jmeyers314 commented on INSTALL.md:
```diff
@@ -22,25 +22,24 @@ Table of Contents:
 
-i) Python (2.6 or 2.7 series), with some additional modules installed
----------------------------------------------------------------------
+i) Python (2.6, 2.7, or 3.4 series), with some additional modules installed
```
I thought we were no longer supporting 2.6. https://github.com/GalSim-developers/GalSim/pull/703#discussion-diff-63630905L364

rmjarvis replied:
I still have a system that has 2.6, so I can still test on it.  It's not a huge burden to support it.  Mostly I just need to keep adding numbers to your `{}`'s in `format` statements.  :)

jmeyers314 replied:
Fair enough, that's pretty easy :).  I went ahead and fixed some remaining numberless `"{}"`s.

jmeyers314 commented:
I only glanced very quickly over the actual code changes, but I can report that this builds and tests fine for me on python 2.7.  Looking into python 3.5 next...

jmeyers314 commented:
Quick question:  Scons apparently doesn't support python 3 yet.  So I have to use python 2 to compile my python 3 program?  Izzat right?

cwwalter commented:
I saw some comments about this from @timj in another LSST context.  He may be able to answer.

> On Jun 18, 2016, at 8:42 AM, Josh Meyers <notifications@github.com> wrote:
> 
> Quick question: Scons apparently doesn't support python 3 yet. So I have to use python 2 to compile my python 3 program? Izzat right?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rmjarvis commented:
I have at least one system where SCons is using python 2.4 still.  But I don't think it's a problem.  Most systems will continue to have a python 2 installation for the foreseeable future.

jmeyers314 commented:
Just noticed that the install.md table of contents was out of sync.  Fixed it up and added links.

timj commented:
@cwwalter Scons is being ported to Python 3 and has dropped 2.4 support. See for example timj/scons#1

jmeyers314 commented:
I guess it just struck me as odd that I needed to be in a python2 environment to compile a python3 application :) (since scons has a usr/bin/env python shebang).

I'm still having trouble compiling for python3 though.  Here's my progress so far:

-  Obtain fresh python 2.7 miniconda.

-  Install scons:
`conda install scons`

-  Get FFTW and TMV from homebrew:
`brew install fftw tmv-cpp`

-  Create python3 conda environment:
`conda create --name py3 python=3`

-  Enter python3 env to install python reqs.  Then exit back to python 2.7 env so scons works.
```
source activate py3
conda install boost numpy astropy pyyaml pandas future
source deactivate
```

-  Fix up paths in libpython and libboost_python.
```
cd /Users/josh/miniconda2/envs/py3/lib`
install_name_tool -id $PWD/libpython3.5m.dylib libpython3.5m.dylib
install_name_tool -id $PWD/libboost_python.dylib libboost_python.dylib
install_name_tool -id $PWD/libboost_python3.dylib libboost_python3.dylib
```

I first tried only fixing up libpython, since the `@rpath`s in the 2.7 version of libboost_python seemed to work okay when I was compiling for 2.7.  The compile for python 3.5 got hung up on the python loadable module check though, until I overwrote the `@rpath`s with the absolute path.

Just out of curiosity, would a better solution be to get the right `@rpath` expansion into the galsim shared object?  I tried using LINKFLAGS='-rpath /Users/josh/miniconda2/envs/py3/lib/'  but that made the TMV link check fail.

-  Install GalSim.  Here's my current scons settings settings:
`scons PYTHON=/Users/josh/miniconda2/envs/py3/bin/python BOOST_DIR=/Users/josh/miniconda2/envs/py3/ EXTRA_LIB_PATH=/Users/josh/miniconda2/envs/py3/lib/`

This appears to get as far as linking the _galsim.so library (command line output below).  Any thoughts on what to try next?

```
scons: Reading SConscript files ...
SCons is version 2.3.0 using python version 2.7.11
Python is from /Users/josh/miniconda2/include/python2.7
Using the following (non-default) scons options:
   PYTHON = /Users/josh/miniconda2/envs/py3/bin/python
   BOOST_DIR = /Users/josh/miniconda2/envs/py3/
   EXTRA_LIB_PATH = /Users/josh/miniconda2/envs/py3/lib/
These can be edited directly in the file gs_scons.conf.
Type scons -h for a full list of available options.
Using python =  /Users/josh/miniconda2/envs/py3/bin/python
Using default PYPREFIX =  /Users/josh/miniconda2/envs/py3/lib/python3.5/site-packages
Using compiler: /usr/bin/g++
compiler version: 4.2.1
Determined that a good number of jobs = 8
Checking for C++ header file fftw3.h... (cached) yes
Checking for correct FFTW linkage... (cached) yes
Checking for boost header files... (cached) yes
Boost version is 1.60.0
Checking for C++ header file TMV.h... (cached) yes
TMV version is 0.72
Using TMV_LINK file: /usr/local/share/tmv/tmv-link
     -L/usr/local/Cellar/tmv-cpp/0.72/lib -ltmv -llapack -lblas
Mac version is 10.10.5
XCode version is 7.2
WARNING: The Apple BLAS library has been found not to be thread safe on
         Mac OS versions 10.7+, even across multiple processes (i.e. not
         just multiple threads in the same process.)  The symptom is that
         `scons tests` may hang when running nosetests using multiple
         processes.
         This seems to have been partially fixed with XCode 5.1, so there
         is a good chance you will not have any problems.  But there are
         still occasional systems that fail when using multithreading with
         programs or modules that link to the BLAS library (such as GalSim).
         If you do have problems, the solution is to recompile TMV with
         the SCons option "WITH_BLAS=false".
Checking for correct TMV linkage... (this may take a little while)
Checking for correct TMV linkage... (cached) yes
Checking if we can build against Python... (cached) yes
Building for python version 3.5
Checking if we can build module using TMV... (cached) yes
Checking if we can build against NumPy... (cached) yes
Checking for PyFITS... (cached) yes
Checking for future... (cached) yes
Checking if we can build against Boost.Python... (cached) yes
Checking if C++ exceptions are propagated up to python... (cached) yes
GalSim version  1.4
scons: done reading SConscript files.
scons: Building targets ...
scons: `lib/libgalsim.1.4.dylib' is up to date.
scons: `lib/libgalsim.dylib' is up to date.
scons: `bin/sizeof_SIFD' is up to date.
g++ -o galsim/_galsim.so -bundle pysrc/.obj/module.os pysrc/.obj/Angle.os pysrc/.obj/Bounds.os pysrc/.obj/Image.os pysrc/.obj/SBProfile.os pysrc/.obj/SBAdd.os pysrc/.obj/SBConvolve.os pysrc/.obj/SBDeconvolve.os pysrc/.obj/SBFourierSqrt.os pysrc/.obj/SBTransform.os pysrc/.obj/SBBox.os pysrc/.obj/SBGaussian.os pysrc/.obj/SBExponential.os pysrc/.obj/SBSersic.os pysrc/.obj/SBMoffat.os pysrc/.obj/SBAiry.os pysrc/.obj/SBShapelet.os pysrc/.obj/SBInterpolatedImage.os pysrc/.obj/SBKolmogorov.os pysrc/.obj/SBSpergel.os pysrc/.obj/Random.os pysrc/.obj/Noise.os pysrc/.obj/HSM.os pysrc/.obj/Integ.os pysrc/.obj/Table.os pysrc/.obj/Interpolant.os pysrc/.obj/CorrelatedNoise.os pysrc/.obj/Bessel.os pysrc/.obj/CDModel.os -Llib -L/Users/josh/miniconda2/envs/py3/lib -L/Users/josh/miniconda2/envs/py3/lib -L/usr/local/Cellar/tmv-cpp/0.72/lib -lboost_python3 -lpython3.5m -ltmv_symband -lfftw3 -ltmv -llapack -lblas -lgalsim
BuildExecutableScript(["bin/galsim"], ["bin/galsim.py"])
BuildExecutableScript(["bin/galsim_yaml"], ["bin/galsim_yaml.py"])
BuildExecutableScript(["bin/galsim_json"], ["bin/galsim_json.py"])
BuildExecutableScript(["bin/galsim_download_cosmos"], ["bin/galsim_download_cosmos.py"])
BuildExecutableScript(["bin/installed_python"], ["bin/installed_python.py"])
Undefined symbols for architecture x86_64:
  "boost::python::objects::function_object(boost::python::objects::py_function const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&)", referenced from:
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*, double), boost::python::default_call_policies, boost::mpl::vector3<void, _object*, double>, mpl_::int_<0> >(void (*)(_object*, double), boost::python::default_call_policies const&, boost::mpl::vector3<void, _object*, double> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<_object* (*)(galsim::AngleUnit&, galsim::AngleUnit const&), boost::python::default_call_policies, boost::mpl::vector3<_object*, galsim::AngleUnit&, galsim::AngleUnit const&>, mpl_::int_<0> >(_object* (*)(galsim::AngleUnit&, galsim::AngleUnit const&), boost::python::default_call_policies const&, boost::mpl::vector3<_object*, galsim::AngleUnit&, galsim::AngleUnit const&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<double (galsim::AngleUnit::*)() const, boost::python::default_call_policies, boost::mpl::vector2<double, galsim::AngleUnit&>, mpl_::int_<0> >(double (galsim::AngleUnit::*)() const, boost::python::default_call_policies const&, boost::mpl::vector2<double, galsim::AngleUnit&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<_object* (*)(galsim::AngleUnit&, double const&), boost::python::default_call_policies, boost::mpl::vector3<_object*, galsim::AngleUnit&, double const&>, mpl_::int_<0> >(_object* (*)(galsim::AngleUnit&, double const&), boost::python::default_call_policies const&, boost::mpl::vector3<_object*, galsim::AngleUnit&, double const&> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<double (*)(galsim::AngleUnit, galsim::AngleUnit), boost::python::default_call_policies, boost::mpl::vector3<double, galsim::AngleUnit, galsim::AngleUnit>, mpl_::int_<0> >(double (*)(galsim::AngleUnit, galsim::AngleUnit), boost::python::default_call_policies const&, boost::mpl::vector3<double, galsim::AngleUnit, galsim::AngleUnit> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*), boost::python::default_call_policies, boost::mpl::vector2<void, _object*>, mpl_::int_<0> >(void (*)(_object*), boost::python::default_call_policies const&, boost::mpl::vector2<void, _object*> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      boost::python::api::object boost::python::detail::make_function_aux<void (*)(_object*, double, galsim::AngleUnit), boost::python::default_call_policies, boost::mpl::vector4<void, _object*, double, galsim::AngleUnit>, mpl_::int_<0> >(void (*)(_object*, double, galsim::AngleUnit), boost::python::default_call_policies const&, boost::mpl::vector4<void, _object*, double, galsim::AngleUnit> const&, std::__1::pair<boost::python::detail::keyword const*, boost::python::detail::keyword const*> const&, mpl_::int_<0>) in Angle.os
      ...
  "boost::python::objects::register_dynamic_id_aux(boost::python::type_info, std::__1::pair<void*, boost::python::type_info> (*)(void*))", referenced from:
      boost::python::class_<galsim::AngleUnit, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Angle.os
      void boost::python::class_<galsim::Angle, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::initialize<boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > >(boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > const&) in Angle.os
      boost::python::class_<galsim::Position<double>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Bounds.os
      void boost::python::class_<galsim::Bounds<double>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::initialize<boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > >(boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > const&) in Bounds.os
      boost::python::class_<galsim::Position<int>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, boost::python::no_init_t) in Bounds.os
      void boost::python::class_<galsim::Bounds<int>, boost::python::detail::not_specified, boost::python::detail::not_specified, boost::python::detail::not_specified>::initialize<boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > >(boost::python::init_base<boost::python::init<mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_, mpl_::void_> > const&) in Bounds.os
      boost::python::class_<galsim::BaseImage<short>, boost::noncopyable_::noncopyable, boost::python::detail::not_specified, boost::python::detail::not_specified>::class_(char const*, char const*, boost::python::no_init_t) in Image.os
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [galsim/_galsim.so] Error 129
scons: building terminated because of errors.
```

rmjarvis commented:
Looks like a c++11 issue.  cf. http://stackoverflow.com/questions/18389799/boost-python-undefined-symbols-error-even-though-the-lib-files-are-linked

Maybe anaconda in their infinite wisdom has a different g++ in their python2 and python3 environments?  Just guessing here.  I can't stand anaconda, so I try to stay as far away from it as possible.  Everything worked fine using fink for python3.

jmeyers314 commented:
>Looks like a c++11 issue.

Alright.  As long as I'm not doing anything obviously silly I can live with abandoning anaconda and sourcing boost from somewhere else.

Not sure if this is relevant, but I see that lib/libgalsim.dylib is linked against /usr/lib/libc++.1.dylib, whereas libboost_python3.dylib is linked against /usr/lib/libstdc++.6.dylib.

rmjarvis commented:
> Not sure if this is relevant, but I see that lib/libgalsim.dylib is linked against /usr/lib/libc++.1.dylib, whereas libboost_python3.dylib is linked against /usr/lib/libstdc++.6.dylib.

Yes.  That looks like the proximate cause of the problem.

rmandelb commented on SConstruct:
```diff
@@ -1145,6 +1145,20 @@ def CheckModuleLibs(config,try_libs,source_file,name,prepend=True):
+    if not result:
+        py_version = ''
+        for v in ['2.7', '2,6', '3.4', # supported versions first
+                  '2.5', '2,4', '3,5', '3.3', '3.2', '3.1', 3.0]:
```
Should the 3.0 be '3.0'?  Should all of these really be in this list?  (2.4?)

rmjarvis replied:
Yes, about `'3.0'`.  And these are just for determining what the version is, not a promise to support it, so I tried to include anything that would be remotely plausible that someone might have for their python distro.  If it doesn't find any of these, then py_version will just be `''`.

rmandelb commented:
Looks like Travis doesn't like this PR.

rmjarvis commented:
Right.  Some more print functions to switch after the merge.

rmandelb commented:
OK - I have a request. Would you be willing to update https://github.com/GalSim-developers/GalSim/wiki/Checklist-for-making-a-pull-request with the few most important things to watch out for with respect to python 3 compatibility, for people like me who don't have a python 3 setup to use for testing?  I tend to check that page to make sure I haven't forgotten anything important, and it would be a good reminder for me if things like print functions (and anything else that comes up really often - maybe nothing?) were mentioned there.

rmandelb commented:
Now Travis has an even weirder problem - I have no idea...

rmjarvis commented:
> Would you be willing to update https://github.com/GalSim-developers/GalSim/wiki/Checklist-for-making-a-pull-request with the few most important things to watch out for with respect to python 3 compatibility, for people like me who don't have a python 3 setup to use for testing? 

Done.  I'll make sure I test with Python 2.6 and 3.4 before any release, but I won't typically test it for every PR.  Similar to how I try to test releases on lots of different compilers, OS's, etc., but sometimes it goes a while before I catch the incompatibilities on master.

> Now Travis has an even weirder problem - I have no idea...

I pushed a dummy commit to trigger another Travis run, and it worked.  I guess Travis just bonks from time to time.  I can relate...

rmandelb commented:
> Done. 

Thanks.

> I guess Travis just bonks from time to time. I can relate...

LOL, yes, me too.

rmjarvis commented:
@jmeyers314 Any luck getting a Python 3 installation going?  

Anyone else have access to a Python 3 set up.  Maybe @timj?  I know you have a Python 3 system. Would you mind trying this branch out to make sure it builds properly for you and passes tests?

jmeyers314 commented:
I think I'm mostly there.  I've got past `scons` && `scons install` without any errors, and I appear to be able to run `python3 test_base.py`, for example.  Is there a natural way to inform nosetests that I want to replace `python` -> `python3` on the command line?

jmeyers314 commented:
Ok.  Figured it out.  Just needed to `pip3 install nosetests`, and then start nosetests with the full path since the new one isn't first in my path.  (Any tips on managing multiple pythons BTW?  My system feels kind of hacked together at the moment :))  I suppose there's no easy way to make `scons tests` do the right thing, but that's okay.  Regardless, I can report that all non-skipped (i.e., non-LSST) tests pass on my hacky python3.5 installation.

rmjarvis commented:
>  I suppose there's no easy way to make scons tests do the right thing, but that's okay. 

You can use `scons tests NOSETESTS='/path/to/python3/nosetests`.

> Any tips on managing multiple pythons BTW?

In my case, I just need to change my PYTHON_PATH to include my local site-packages for 3.4 and switch the fink boost installation.  I have a script called `setup_python3` that does this and `setup_python2` that goes back.  

In the GalSim directory, I keep around lots of `gs_scons.*` files for different configurations.  One of these is `gs_scons.py3` which I copy to `gs_scons.conf` when I'm doing python3 stuff.

> Regardless, I can report that all non-skipped (i.e., non-LSST) tests pass on my hacky python3.5 installation.

Great!  Should we add 3.5 to the list of officially supported pythons then?  i.e. are you willing to make sure this works for you each time we get to a release?  (No need to do this for every PR.)

jmeyers314 commented:
>You can use scons tests NOSETESTS='/path/to/python3/nosetests'.

Ah, great!  Works perfectly.

>Should we add 3.5 to the list of officially supported pythons then? i.e. are you willing to make sure this works for you each time we get to a release? (No need to do this for every PR.)

Sure.  That's pretty infrequent, so not too arduous.

rmandelb commented:
@rmjarvis - I went over the CHANGELOG and INSTALL updates carefully.  Once I got into the code, my eyes started to glaze over given the nature of the changes.  Personally, I am comfortable with saying that you haven't broken either of my two installations (Mac and Linux), that I can run unit tests and demos as before, and I am not going to read the code but am happy with this PR to be merged as soon as everyone else's installation issues are settled.

However - if there are more interesting code changes hidden somewhere in there that you would like me to look at, please let me know and i will check them out.

rmjarvis commented:
No, there's nothing particularly interesting in there.  Like I said, mostly I just followed the advice on the cheat sheet about how to make things compatible with both python 2 and python 3.  I think the main thing is that installation and tests work, so let's plan to merge Monday unless anyone objects.

jmeyers314 commented:
I just noticed that demo13.py is failing for me under python2.7 with the message:

```
Traceback (most recent call last):
  File "demo13.py", line 473, in <module>
    main(sys.argv)
  File "demo13.py", line 462, in main
    final_image.write(out_filename)
  File "/Users/josh/src/GalSim/galsim/fits.py", line 550, in write
    image.wcs.writeToFitsHeader(hdu.header, image.bounds)
  File "/Users/josh/src/GalSim/galsim/wcs.py", line 547, in writeToFitsHeader
    for key in self.header:
TypeError: iter() returned non-iterator of type 'instancemethod'
```

which appears to be related to [this line](https://github.com/GalSim-developers/GalSim/blob/%23534/galsim/wcs.py#L547) in wcs.py.

I also noticed that fits.FitsHeader has two definitions of `__len__`.

May be this is due to most recent version of astropy (v1.2)?

I'll check these on python3.5 too.

jmeyers314 commented:
There was a somewhat more mundane error for demo13.py in python3.5 related to floating point division.  Once that's fixed though, I run into the same error as in python2.7: `TypeError: iter() returned non-iterator of type 'instance method'`  (python3 astropy is v1.2.1).

I also updated `psf_wf_movie.py` so that it works in both python2.7 and python3.5.

rmjarvis commented:
> I just noticed that demo13.py is failing for me under python2.7 with the message:

Thanks for catching that.  That was an error in the FitsHeader `__iter__` method, which I fixed.  I don't always remember to run 12 and 13, since they aren't in check_yaml.  So I also added them there.  There's no diff to check, since there aren't yaml versions of these.  But at least it will check that they run to completion.

> I also noticed that fits.FitsHeader has two definitions of __len__.

Right.  That may have been there for a while.  The second one is the correct one, which overwrites the first one when parsing the file, so it wasn't an error.  But we still shouldn't have two.  I removed the first one.

jmeyers314 commented:
@rmjarvis everything works for me now!  I vote merge as scheduled Monday morning.
